src/main/java/bio/terra/workspace/service/datareference/DataReferenceService.java

```diff
@@ -0,0 +1,54 @@
+package bio.terra.workspace.service.datareference;
+
+import bio.terra.workspace.common.exception.SamUnauthorizedException;
+import bio.terra.workspace.common.exception.WorkspaceNotFoundException;
+import bio.terra.workspace.common.utils.SamUtils;
+import bio.terra.workspace.db.DataReferenceDao;
+import bio.terra.workspace.db.WorkspaceDao;
+import bio.terra.workspace.generated.model.DataReferenceList;
+import bio.terra.workspace.generated.model.FilterControlledEnum;
+import bio.terra.workspace.service.iam.AuthenticatedUserRequest;
+import bio.terra.workspace.service.iam.SamService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+public class DataReferenceService {
+
+  private SamService samService;
+  private DataReferenceDao dataReferenceDao;
+  private WorkspaceDao workspaceDao;
+
+  @Autowired
+  public DataReferenceService(
+      SamService samService, DataReferenceDao dataReferenceDao, WorkspaceDao workspaceDao) {
+    this.samService = samService;
+    this.dataReferenceDao = dataReferenceDao;
+    this.workspaceDao = workspaceDao;
+  }
+
+  public DataReferenceList enumerateDataReferences(
+      String workspaceId,
+      int offset,
+      int limit,
+      FilterControlledEnum filterControlled,
+      AuthenticatedUserRequest userReq) {
+    if (!samService.isAuthorized(
+        userReq.getRequiredToken(),
+        SamUtils.SAM_WORKSPACE_RESOURCE,
+        workspaceId,
+        SamUtils.SAM_WORKSPACE_READ_ACTION)) {
+      throw new SamUnauthorizedException(
+          "User " + userReq.getEmail() + " is not authorized to read workspace " + workspaceId);
+    }
+    // TODO: do we want to give users a different error when the specified workspace doesn't exist?
+    //  It requires an additional DB lookup (as an empty enumerate doesn't tell you whether
+    //  or not a workspace exists), but it's a nicer user experience for catching small mistakes.
+    if (!workspaceDao.workspaceExists(workspaceId)) {
+      throw new WorkspaceNotFoundException("No workspace exists with ID " + workspaceId);
+    }
```

Comment:
if the workspace does not exist, then won't the previous Sam call return an error? I may be wrong about that.

We should not allow users to fish for the existence/nonexistence of a workspace.
* if the workspace exists and user is allowed to see it, all good
* if and only if the user has permission to list all workspaces (in a given context, such as if they are a folder owner), but the workspace does not exist, then we can/should return a 404
* if the user does not have permission to list all workspaces, or the user is unauthorized for this workspace, then we return 403 with a generic message of "user is not authorized or workspace does not exist"

Reply:
Checking the documentation, it looks like Sam returns false if a user is unauthorized or the resource doesn't exist so yep, it'll throw an error. Took out the "workspace exists" check and updated the error message.

Since Sam doesn't tell us whether the authorization failed or the resource just doesn't exist, I'm not sure how to check for your second case here without adding more business logic, and I don't think the convenience is worth that.

Reply:
Also removed the missing workspace test, since that's now not distinct from the "unauthorized" case.